csfieldguide/static/interactives/cfg-equation-builder/js/cfg-equation-builder.js

```diff
@@ -0,0 +1,549 @@
+  }
+  return terminals;
+}
+
+/******************************************************************************/
+// FUNCTIONS FOR PREPARING THE INTERACTIVE //
+/******************************************************************************/
+
+/**
+ * Fills the user-facing table with the given grammar productions.
+ * If the productions are sufficiently short, they are displayed in 2 columns
+ */
+function fillProductionsWindow(productions) {
+  var keys = Object.keys(productions)
+  var describedProductions = [];
+  for (i=0; i<keys.length; i++) {
+    describedProductions = describedProductions.concat(describeAndReduceProductions(keys[i], productions[keys[i]]));
+  }
+
+  var use2Cols = describedProductions.length > 1;
+  for (var x=0; x<describedProductions.length; x++) {
+    // 14 is the length of &#8594 plus 8 characters
+    if (describedProductions[x].length > 14) {
+      use2Cols = false;
+      break;
+    }
+  }
+
+  var $table = $('#productions-table')
+  var table = "";
+  if (use2Cols) {
+    table += (`<tr><th colspan="2"><h3>${gettext("Productions:")}</h3></th></tr>`)
+    var half = Math.ceil(describedProductions.length / 2);
+    for (var i=0; i<half; i++) {
+      if (half + i < describedProductions.length) {
+        table += (`<tr><td>${describedProductions[i]}</td><td>${describedProductions[half + i]}</td></tr>`)
+      } else {
+        table += (`<tr><td>${describedProductions[i]}</td><td></td></tr>`)
+      }
+    }
+  } else {
+    table += (`<tr><th><h3>${gettext("Productions:")}</h3></th></tr>`)
+    for (var j=0; j<describedProductions.length; j++) {
+      table += (`<tr><td>${describedProductions[j]}</td></tr>`)
+    }
+  }
+  $table.html(table);
+}
+
+/**
+ * Returns a list of strings, each describing productions from a given nonterminal
+ * 
+ * If replacements is an incremental list of integers they are reduced appropriately
+ */
+function describeAndReduceProductions(nonterminal, replacements) {
+  var isCompressable = true;
+  if (typeof(replacements[0]) == 'number' && replacements.length > 1) {
+    for (var i=1; i<replacements.length; i++) {
+      if (replacements[i] != replacements[i-1] + 1) {
+        isCompressable = false;
+        break;
+      }
+    }
+  } else {
+    isCompressable = false;
+  }
+
+  if (isCompressable) {
+    return [`${nonterminal} &#8594 ${replacements[0]}-${replacements[replacements.length - 1]}`]
+  }
+  var returnList = [];
+  for (i=0; i<replacements.length; i++) {
+    returnList.push(describeProduction(nonterminal, replacements[i]));
+  }
+  return returnList;
+}
+
+/******************************************************************************/
+// FUNCTIONS FOR INTERACTIVE LOGIC //
+/******************************************************************************/
+
+/**
+ * Sets the equation generator to be of the given type
+ */
+function setGenerator(type) {
+  $('#generate-button').html($('#set-g-' + type).html());
```

Comment:
Could we maybe save `$('#set-g-' + type)` to a variable to make this a bit more readable?

Or even setting `$('#set-g-' + type).html()` and `$('#set-g-' + type).attr('g-type')` to variables. I know they only get used once but the readability would be better.